uvicorn/server.py

```diff
@@ -270,6 +271,11 @@ async def shutdown(self, sockets: list[socket.socket] | None = None) -> None:
         for sock in sockets or []:
             sock.close()
 
+        if callable(self.config.before_graceful_exit_hook):
+            f = self.config.before_graceful_exit_hook()
+            if inspect.isawaitable(f):
+                await f
```

Comment:
Is it possible to place this code prior to the point where new connections are no longer accepted (#Stop accepting new connections)? 

This could help in resolving the matter we have mentioned in #2257. Additionally, since the topic at hand is related to "before graceful exit," it would be good to execute the code beforehand, before the acceptance of any further connections is stopped.

WDYT?

Reply:
I have no objection to this, for me these two positions are the same.

Reply:
@adriangb @Kludex , what are you thoughts on this?